File: Modelo_PCA.jl
using MultivariateStats
using Plots
using Random
using Statistics

# 1. Gerar um conjunto de dados sintético
Random.seed!(42)  # Para reprodutibilidade
n_samples = 300
n_features = 5
n_centers = 3

# Função para criar blobs de dados
function make_blobs(n_samples, n_features, n_centers)
    centers = randn(n_centers, n_features) * 10
    X = vcat([randn(n_samples ÷ n_centers, n_features) .+ centers[i, :] for i in 1:n_centers]...)
    y = repeat(1:n_centers, inner=n_samples ÷ n_centers)
    return X, y
end

X, y = make_blobs(n_samples, n_features, n_centers)

# 2. Aplicar PCA para reduzir a dimensionalidade para 2 componentes principais
pca_model = fit(PCA, X; maxoutdim=2)
X_pca = transform(pca_model, X)

# 3. Visualizar os resultados
scatter(X_pca[:, 1], X_pca[:, 2], group=y, legend=:topright,
        xlabel="Componente Principal 1", ylabel="Componente Principal 2",
        title="Projeção PCA dos Dados", palette=:viridis)

# 4. Explicar a variância acumulada
explained_variance = principalvars(pca_model)
total_variance_explained = sum(explained_variance[1:2]) / sum(explained_variance)
println("Variância explicada pelo primeiro componente principal: $(explained_variance[1] / sum(explained_variance) * 100):.2f%")
println("Variância explicada pelo segundo componente principal: $(explained_variance[2] / sum(explained_variance) * 100):.2f%")
println("Variância total explicada pelos dois primeiros componentes principais: $(total_variance_explained * 100):.2f%")

File: Modelo_PCA.py
import numpy as np
import matplotlib.pyplot as plt
from sklearn.decomposition import PCA
from sklearn.datasets import make_blobs

# 1. Gerar um conjunto de dados sintético
X, y = make_blobs(n_samples=300, n_features=5, centers=3, random_state=42)

# 2. Aplicar PCA para reduzir a dimensionalidade para 2 componentes principais
pca = PCA(n_components=2)
X_pca = pca.fit_transform(X)

# 3. Visualizar os resultados
plt.figure(figsize=(8, 6))
plt.scatter(X_pca[:, 0], X_pca[:, 1], c=y, cmap='viridis', edgecolor='k', s=100)
plt.xlabel('Componente Principal 1')
plt.ylabel('Componente Principal 2')
plt.title('Projeção PCA dos Dados')
plt.colorbar(label='Classe')
plt.show()

# 4. Explicar a variância acumulada
explained_variance = pca.explained_variance_ratio_
print(f'Variância explicada pelo primeiro componente principal: {explained_variance[0]:.2f}')
print(f'Variância explicada pelo segundo componente principal: {explained_variance[1]:.2f}')
print(f'Variância total explicada pelos dois primeiros componentes principais: {explained_variance.sum():.2f}')

